homeassistant/components/overkiz/__init__.py

```diff
@@ -144,3 +147,45 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+async def _async_migrate_entries(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> bool:
+    """Migrate old entry."""
+    entity_registry = er.async_get(hass)
+
+    @callback
+    def update_unique_id(entry: er.RegistryEntry) -> dict[str, str] | None:
+        # Python 3.11 treats (str, Enum) and StrEnum in a different way
+        # Since pyOverkiz switched to StrEnum, we need to rewrite the unique ids once to the new style
+        if (key := entry.unique_id.split("-")[-1]).startswith("OverkizState"):
+            state = key.split(".")[1]
+            new_unique_id = entry.unique_id.replace(key, OverkizState[state])
+
+            LOGGER.debug(
+                "Migrating entity '%s' unique_id from '%s' to '%s'",
+                entry.entity_id,
+                entry.unique_id,
+                new_unique_id,
+            )
+
+            if existing_entity_id := entity_registry.async_get_entity_id(
+                entry.domain, entry.platform, new_unique_id
+            ):
+                LOGGER.debug(
+                    "Cannot migrate to unique_id '%s', already exists for '%s'",
+                    new_unique_id,
+                    existing_entity_id,
+                )
+                return None
```

Comment:
Since this entity stopped working in HA Core 2023.3 until now, with the exception of again working in 2023.9.0, I suggest instead of returning `None` here we deem this entity to be a duplicate and delete it.

This will require changing `entity_registry.async_migrate_entries` to allow deleting entities though by iterating over a temporary list.

Reply:
Thanks! I was already working on something similar, where I store all ids that are duplicated in another list, and loop later through this list for removal. But it would be cleaner to have such functionality in `entity_registry.async_migrate_entries`.

Reply:
Not sure if I can get this in (with tests) on time for beta. It raises an error on removal of the entities, while the `3541212-OverkizState.CORE_DISCRETE_RSSI_LEVEL` entity is present in my `storage/core.entity_registry`. This sensor is `disabled_by: integration`, but not sure if that can cause an issue.

```
2023-09-30 10:27:26.115 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry [x] for overkiz
Traceback (most recent call last):
  File "/workspaces/home-assistant-core/homeassistant/config_entries.py", line 399, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/home-assistant-core/homeassistant/components/overkiz/__init__.py", line 59, in async_setup_entry
    await _async_migrate_entries(hass, entry)
  File "/workspaces/home-assistant-core/homeassistant/components/overkiz/__init__.py", line 191, in _async_migrate_entries
    await er.async_migrate_entries(hass, config_entry.entry_id, update_unique_id)
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity_registry.py", line 1339, in async_migrate_entries
    updates = entry_callback(entry)
              ^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/home-assistant-core/homeassistant/components/overkiz/__init__.py", line 181, in update_unique_id
    entity_registry.async_remove(entry.unique_id)
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity_registry.py", line 708, in async_remove
    entity = self.entities.pop(entity_id)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen _collections_abc>", line 912, in pop
  File "/usr/local/lib/python3.11/collections/__init__.py", line 1124, in __getitem__
    raise KeyError(key)
KeyError: 'io://xxxx-xxxx-xxxx/3541212-OverkizState.CORE_DISCRETE_RSSI_LEVEL'
```